### PR TITLE
Rewrite of the dashboard reconciliation

### PR DESCRIPTION
**Pull request by kaguillera**
This is a rewrite of the reconciliation report in the dashboard directory (https://github.com/gratipay/finances/issues/21) It still needs to be tested on a snapshot of actual data to verify it works properly in production. Any advice on how I can recreate a snapshot of the database to do proper testing

### CODE REVIEW

**Comment by whit537** [comment]
Rebased on master at c213b9af9337e79026bf0fed89b098d0c44ebad7. Previous head was fdd18d66c0725b559041619738e5804773b4060b.

File: www/dashboard/reconciliation.spt [hunk 1]
```diff
@@ -68,45 +118,47 @@ fmt = lambda x: "{:,.2f}".format(x)
...
-            </tr>
-            {% for month, payins, payouts, balance, fees in by_month %}
-            <tr>
-                <td>{{ str(month)[:7] }}</td>
-                <td>{{ fmt(payins) }}</td>
-                <td>{{ fmt(payouts) }}</td>
-                <td>{{ fmt(balance) }}</td>
-                <td>{{ fmt(fees) }}</td>
-            </tr>
+            {% for month, recs in by_month|dictsort %}
```
Comment (whit537) [inline www/dashboard/reconciliation.spt]: Nice. Is this a Jinja2 thing?

Comment (whit537) [inline www/dashboard/reconciliation.spt]: [Yes](http://jinja.pocoo.org/docs/dev/templates/#dictsort).

**Comment by whit537** [comment]
I see `balanced-cc` in 2012-07 as the first month and network. We should have *all* exchanges on this report, regardless of whether they have a route or not. A subhead for "no route" would actually be helpful in working through https://github.com/gratipay/gratipay.com/pull/3912.

**Comment by whit537** [comment]
Maybe http://stackoverflow.com/questions/2054235/why-doesnt-postgres-group-by-null-select-counts?

**Comment by kaguillera** [comment]
So these are the steps I think I have to make to show the null values in the dashboard reconciliation page:

- [ ] Create and branch.sql in local /sql/ folder to alter exchanges and payment_net
     - [x] Alter exchange_status and payment_net by adding ‘ unknown’ type to enum
     - [ ] Insert records for ‘unknown’  network in exchange_route table
     - [ ] Update exchanges records with exchange_route ids pointing to ‘unknown’ network records for that participants
     - [x] Update all null exchange_status to ‘ unknown'
     - [ ] Alter exchanges table and set status to not null
     - [ ] Alter exchange_routes table and set route to not null
- [ ] run and test changes made locally
- [ ] commit changes once satisfied it is working properly

**Comment by kaguillera** [comment]
This last commit is the changes made to database in order to capture the NULL values for route in the exchanges table that are not show on the reconciliation page. I tested it here but need real data as before so please review @whit537 as before... scratch I still have some worked based on the failed travis. I will look at it over the weekend.

**Comment by whit537** [comment]
Redid 35fc223c4eb36fa4a888e9c8bdf80011abd37b85 as c370b0df79ad86d91b83707b08fbc6d9825bd38b to uncomment SQL per IRL convo w/ @kaguillera.

**Comment by nashe** [comment]
@whit537 : is the reconciliation page supposed to be public, since it's in the dashboard? Just to be sure.

**Comment by whit537** [comment]
Yeah, if I remember right, earlier in the PR we had two sections on the reconciliation dashboard, one aggregate and one more detailed. The detailed one we wanted to protect, but I'm pretty sure it's gone now. We're okay with the aggregate numbers being public.

**Comment by nashe** [comment]
Ok, thanks for the reply ;-)

**Comment by kaguillera** [comment]
Finally started back working on this and realize(Trinidad/UK spelling) the reason for the error on Travis. 
In `fake_data.py` the `fake_exchange` function does not associate the exchange with an `exchange_route`. Thus to solve it I have to make the exchange_routes for the fake participants and then associate (foreign_key reference)  them the fake exchanges that are created for that participant. So here goes...

**Comment by kaguillera** [comment]
Now this is finally up for review since travis is passing. ping @whit537

**Comment by whit537** [comment]
@kaguillera Ready for a rebase post-#3596, ya?

**Comment by kaguillera** [comment]
Let me just verify this and get back to you...

**Comment by kaguillera** [comment]
Finally ready for review and merging.

**Comment by whit537** [comment]
Need another rebase. 🐑

**Comment by whit537** [comment]
Talking through git wonkiness here. Trying to understand where the infamous "@kaguillera has added 250+ commits" is coming from. Planning to pick up tomorrow ...

**Comment by kaguillera** [comment]
🎉  No more conflicts 🎉  ready for review again

File: www/dashboard/reconciliation_new.spt [hunk 2]
```diff
@@ -0,0 +1,163 @@
+import calendar
```
Comment (whit537) [inline www/dashboard/reconciliation_new.spt]: What's up with this `reconciliation_new.spt` file? Is it supposed to replace `reconciliation.spt`?

Comment (kaguillera) [inline www/dashboard/reconciliation_new.spt]: `reconciliation_new.spt` is a scratch file that I need to (and going to now) remove. Originally it was going to be a separate page but decided against it I actually thought that I had removed it.

**Comment by whit537** [comment]
I had thought to get this merged and out of the queue before diving into https://github.com/gratipay/inside.gratipay.com/issues/432 and https://github.com/gratipay/gratipay.com/issues/4148, but with the significant changes to the `exchanges` table (not just changes to an admin dashboard) I think we need to be more careful. Like, will the `not null` on `exchanges.route` break something because we currently still leak nulls into there during payday, or something? Punting ...

**Comment by kaguillera** [comment]
I had considered the nulls on the exchange table when I did this pulls request but I can always review this when things settle down a little and what we are doing becomes a little more concrete

**Comment by whit537** [comment]
Looking in the production database, the last `NULL` route I see in the `exchanges` table is on October 9, 2015, which was [The Great Balanced Refund of Twenty Fifteen](https://github.com/gratipay/gratipay.com/pull/3810). I do find https://github.com/gratipay/gratipay.com/issues/3822, which suggests that we haven't satisfied ourselves that there's not _some_ leak still outstanding. Merging this PR and enforcing non-`NULL` will be a good way to flush out those bugs, if they're there.

**Comment by whit537** [comment]
Rebased on `master`. Previous head was bd7fd89d10e477c5779eff8af692a97ebc36acd7.

**Comment by kaguillera** [comment]
So what do you want me to do to close this off

**Comment by whit537** [comment]
Let me try to take a look after lunch. If I don't find any changes needed then I will work on deploying!

**Comment by whit537** [comment]
Rebased. Previously add5c33923a61282c35e0ba66f07576ea9596b72.

**Comment by whit537** [comment]
```
$ psql gratipay-bak < sql/branch.sql 
Null display is "¤".
Line style is unicode.
Border style is 2.
ALTER TYPE
ALTER TYPE
UPDATE 7930
ALTER TABLE
INSERT 0 1626
UPDATE 84308
ALTER TABLE
$
```

**Comment by whit537** [comment]
> the last `NULL` route I see in the `exchanges` table

That's irrelevant here, though. This PR alters `status`, not `route`. The last `NULL` `status` I find is even earlier, 2014-09-19.

**Comment by whit537** [comment]
Hmm ... Travis failures—post-rebase?

**Comment by whit537** [comment]
> That's irrelevant here, though.

Just kidding! ☺️ It alters `status` *and* `route`. 👍

**Comment by whit537** [comment]
Weird. Green again! O.O

**Comment by kaguillera** [comment]
Rebased

**Comment by kaguillera** [comment]
Ready again...what is the verdict...can we merge?

File: sql/branch.sql [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+UPDATE exchanges SET status = 'unknown' WHERE status IS NULL;
+
+-- Alter the exchanges table to ensure that no more NULL values are entered
+ALTER TABLE exchanges ALTER COLUMN status SET NOT NULL;
+
+-- Insert records for ‘unknown’ (previously NULL in exchanges table
+-- network in exchange_route table
+INSERT INTO  exchange_routes (participant, network, address, error)
+     (
+       SELECT DISTINCT participants.id, 'unknown'::payment_net, 'None', 'None'
```
Comment (whit537) [inline sql/branch.sql]: Goofy to use the string "None" here. I think the empty string might be a better choice?

Comment (whit537) [inline sql/branch.sql]: There's a uniqueness constraint on `(participant, network, address)`, which is why non-unique `(network, address)` is fine.

```sql
gratipay-bak=# \d exchange_routes
                                Table "public.exchange_routes"
┌─────────────┬───────────────┬──────────────────────────────────────────────────────────────┐
│   Column    │     Type      │                          Modifiers                           │
├─────────────┼───────────────┼──────────────────────────────────────────────────────────────┤
│ id          │ integer       │ not null default nextval('exchange_routes_id_seq'::regclass) │
│ participant │ bigint        │ not null                                                     │
│ network     │ payment_net   │ not null                                                     │
│ address     │ text          │ not null                                                     │
│ error       │ text          │ not null                                                     │
│ fee_cap     │ numeric(35,2) │                                                              │
└─────────────┴───────────────┴──────────────────────────────────────────────────────────────┘
Indexes:
    "exchange_routes_pkey" PRIMARY KEY, btree (id)
    "exchange_routes_participant_network_address_key" UNIQUE CONSTRAINT, btree (participant, network, address)
Check constraints:
    "exchange_routes_address_check" CHECK (address <> ''::text)
Foreign-key constraints:
    "exchange_routes_participant_fkey" FOREIGN KEY (participant) REFERENCES participants(id)
Referenced by:
    TABLE "exchanges" CONSTRAINT "exchanges_route_fkey" FOREIGN KEY (route) REFERENCES exchange_routes(id)
```

Comment (kaguillera) [inline sql/branch.sql]: Soooo...if we replace the `None` for address and error will that be sufficient or do you still have some concerns

Comment (whit537) [inline sql/branch.sql]: Yeah, let's switch to `n/a` for address and the empty string for error.

Comment (whit537) [inline sql/branch.sql]: Done in bac13defde718a3755ac24dc35c3b260d84f644c.

**Comment by whit537** [comment]
Alright, I'm good here. I'm satisfied that we don't have code inserting *new* `NULL` `route`s or `status`es, and I'm seeing this do what's intended in terms of converting existing `NULL`s to `unknown`. Deploying! 👍

**Pull request merged**